### PR TITLE
gh-82565: Fix a possible assertion failure in _pickle

### PR DESCRIPTION
**Pull request by ZackerySpytz**
In _Unpickler_SetInputStream(), _PyObject_LookupAttrId() is called
three times in a row without any error handling.  If an exception
occurs during the first or second call, _PyObject_LookupAttrId()
will be called with a live exception.

https://bugs.python.org/issue38384

* Issue: gh-82565

### CODE REVIEW

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Please add a test.

**Comment by serhiy-storchaka** [comment]
Minimal reproducer (but you can add `readline` and `readinto`):
```python
class F:
    @property
    def read(self):
        1/0

import pickle
pickle.load(F())
```

**Comment by taleinat** [comment]
@ZackerySpytz Ping?

**Comment by taleinat** [comment]
Hi @ZackerySpytz! This is looking good!

Make sure to also add a NEWS entry (you can use [blurb-it](https://blurb-it.herokuapp.com/)).

File: Lib/test/pickletester.py [hunk 1]
```diff
@@ -3136,6 +3136,18 @@ def __init__(self): pass
         self.assertRaises(pickle.PicklingError, BadPickler().dump, 0)
         self.assertRaises(pickle.UnpicklingError, BadUnpickler().load)
 
+    def test_load_read_exception(self):
```
Comment (serhiy-storchaka) [inline Lib/test/pickletester.py]: `readinto` will not be tested if `read` raises an error. You need six tests for classes which do not have one of these attributes or raise error for it but have other two attributes.

File: Modules/_pickle.c [hunk 2]
```diff
@@ -1609,22 +1609,27 @@ _Unpickler_SetInputStream(UnpicklerObject *self, PyObject *file)
...
-        }
-        Py_CLEAR(self->read);
-        Py_CLEAR(self->readinto);
-        Py_CLEAR(self->readline);
-        Py_CLEAR(self->peek);
-        return -1;
+    if (_PyObject_LookupAttrId(file, &PyId_read, &self->read) <= 0) {
+        goto error;
+    }
+    if (_PyObject_LookupAttrId(file, &PyId_readinto, &self->readinto) <= 0) {
```
Comment (serhiy-storchaka) [inline Modules/_pickle.c]: `readinto` is optional now.

Ifs for `read` and `readline` can be merged.

**Review by bumper-furkan-onder — APPROVED** [review]

**Review by furkanonder — APPROVED** [review]

**Comment by furkanonder** [comment]
LGTM. @serhiy-storchaka

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM.

Please add few more test cases:

* no `read`, no `readline`
* `read = bad_property`, no `readline`
* `readline = bad_property`, no `read`

**Comment by serhiy-storchaka** [comment]
Sorry, I forget to merge this PR at its time, so the original issue was fixed in #105667.

But the tests may still be useful.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**